packages/components/src/number-control/index.tsx

```diff
@@ -107,36 +88,59 @@ function UnforwardedNumberControl(
 		( state, action ) => {
 			const nextState = { ...state };
 
-			const { type, payload } = action;
-			const event = payload.event;
 			const currentValue = nextState.value;
 
 			/**
 			 * Handles custom UP and DOWN Keyboard events
 			 */
 			if (
-				type === inputControlActionTypes.PRESS_UP ||
-				type === inputControlActionTypes.PRESS_DOWN
+				action.type === inputControlActionTypes.PRESS_UP ||
+				action.type === inputControlActionTypes.PRESS_DOWN
 			) {
-				// @ts-expect-error TODO: Resolve discrepancy between `value` types in InputControl based components
-				nextState.value = spinValue(
-					currentValue,
-					type === inputControlActionTypes.PRESS_UP ? 'up' : 'down',
-					event as KeyboardEvent | undefined
+				const actionEvent = (
+					action as inputControlActionTypes.KeyEventAction
+				 ).payload.event;
+				const valueToSpin = isValueEmpty( currentValue )
+					? baseValue
+					: ensureNumber( currentValue );
+
+				nextState.value = ensureString(
+					spinValue( {
+						value: valueToSpin,
+						direction:
+							action.type === inputControlActionTypes.PRESS_UP
+								? 'up'
+								: 'down',
+						event: actionEvent as KeyboardEvent,
+						isShiftStepEnabled,
+						shiftStep: shiftStepAsNumber,
+						baseStep,
+						isStepAny,
+						max,
+						min,
+					} )
 				);
 			}
 
 			/**
 			 * Handles drag to update events
 			 */
-			if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
-				// @ts-expect-error TODO: See if reducer actions can be typed better
```

Comment:
In order to allow better action typing, I removed the destructuring of the `action` argument from the top of the function. This allows the code to typecast `action` to the specific action inside each `if` clause, which makes TypeScript happy!